### PR TITLE
feat(finetune): support messages datasets

### LINKED ISSUE #822: [REQ] add messages dataset support
Parent: #214

Depends on #820.

## Summary

Add support for OpenAI/Hugging Face-style `messages` datasets using the model tokenizer's existing chat template. The initial implementation uses full-sequence loss.

```yaml
datasets:
  - source: organization/chat-data
    type: messages
```

Expected record:

```json
{
  "messages": [
    {"role": "system", "content": "You are concise."},
    {"role": "user", "content": "What is a container image?"},
    {"role": "assistant", "content": "An immutable application package."}
  ]
}
```

## Motivation

`messages` is the common modern schema for multi-turn chat data. Although upstream TRL accepts conversational datasets, the active Unsloth-patched SFT preprocessing path does not consume a raw `messages` column without a formatting function. AIKit must explicitly validate and render messages before training.

## Proposed changes

- Add a `messages` dataset type constant and runtime dispatch.
- Validate each row:
  - Non-empty `messages` list.
  - String `role` and `content` fields.
  - Initially allow only `system`, `user`, and `assistant`.
  - Require at least one assistant turn.
  - Require the final turn to be assistant for the initial contract.
  - Reject structured/multimodal content.
- Require a usable tokenizer chat template; fail clearly for base models without one.
- Render with:

```python
tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=False,
)
```

- Store rendered output as canonical `text` for the active Unsloth SFT path.
- Use full-sequence loss across system, user, and assistant tokens.
- Do not rely on native TRL messages preprocessing.
- Avoid adding special tokens already emitted by the template.
- Include source and row context in errors without exposing sensitive URL query strings.
- Add fixtures, documentation, and a one-step GPU smoke test using an instruct model with a known chat template.

## Acceptance criteria

- `type: messages` is accepted by Go and Python dispatch.
- Valid single- and multi-turn conversations render through the tokenizer template.
- `add_generation_prompt` is always `False` during training rendering.
- The trainer receives rendered `text`, not an unprocessed `messages` column.
- Models without a chat template fail with an actionable error.
- Malformed messages, unsupported roles, or non-string content fail clearly.
- Rendered records contain no AIKit-introduced duplicate BOS/EOS.
- Full-sequence labels supervise all rendered tokens.
- No assistant-/response-only behavior is claimed or silently applied.
- Existing Alpaca and prompt-completion behavior remains unchanged.
- A one-step GPU run exports a valid GGUF and produces a non-empty inference response.

## Tests

- Go validation tests for `type: messages`.
- Python tests for valid single/multi-turn records, optional system messages, malformed fields, unsupported roles, no assistant turn, user-final conversations, missing chat templates, and rendering failures.
- Assert `tokenize=False` and `add_generation_prompt=False`.
- Assert the Alpaca formatter is bypassed.
- Exact-lock preprocessing test proving the Unsloth-patched trainer consumes rendered text.
- Add:
  - `test/fixtures/messages-smoke.jsonl`
  - `test/aikitfile-unsloth-messages-smoke.yaml`
- Run `make test`, `make test-python`, `make lint`, and the GPU fine-tune/export/inference smoke test.

## Out of scope

- Response-only or assistant-only loss.
- ShareGPT normalization.
- Tool calls or `tool` roles.
- Structured or multimodal message content.
- User-provided Jinja templates.
- Multiple datasets.
- DPO, GRPO, vision, or audio.

### PR DESCRIPTION
**Pull request by sozercan**
Stacked on #830. This PR targets `issue-821` and should be merged after #830.

## Summary

- add text-only OpenAI/Hugging Face `messages` datasets with full-sequence SFT
- validate canonical roles/content while allowing benign top-level metadata and rejecting tool/template/multimodal semantics
- require a usable deterministic tokenizer chat template and reject `strftime_now` templates
- render one conversation at a time with `tokenize=false` and `add_generation_prompt=false`
- compare canonical chat-template token IDs with the locked Unsloth rendered-text path using bounded validation
- verify BFD packing boundaries, duplicate/reordered records, and actual full-sequence labels before training
- add Qwen2.5 messages smoke coverage, docs, and fixtures

No response-only, ShareGPT, tool-calling, or multimodal behavior is included.

## Tests

- `make test`
- `make test-python` (88 tests)
- `make check-unsloth-lock` with uv 0.12.1
- CI-pinned `golangci-lint` 2.11.2
- Actionlint and fixture validation
- `go mod tidy` clean

The GPU `messages-smoke` job will be run on the completed top-of-stack branch.

Closes #822

### CODE REVIEW

File: pkg/utils/const.go [hunk 1]
```diff
@@ -14,9 +14,26 @@ const (
 	TargetUnsloth = "unsloth"
 
 	DatasetAlpaca           = "alpaca"
+	DatasetMessages         = "messages"
+	DatasetPreference       = "preference"
 	DatasetPromptCompletion = "prompt-completion"
+	DatasetShareGPT         = "sharegpt"
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/utils/const.go:20]: This expanded diff is an artifact of the final stacked merge, not an unreviewed scope expansion in the original PR. Each downstream feature remained in its own PR with independent review, unit tests, documentation, and live GPU validation (#831 through #835). During landing, the already-reviewed child branches were merged upward through their feature bases so the stack could be merged into `main`; that made the lower PR's final branch contain the downstream commits. No additional implementation change is needed for this scope-only thread.

File: pkg/build/build.go [hunk 2]
```diff
@@ -428,27 +436,96 @@ func validateFinetuneConfig(c *config.FineTuneConfig) error {
...
+	if len(c.Datasets) == 0 {
+		return errors.New("no datasets defined")
 	}
 
-	for _, dataset := range c.Datasets {
+	for datasetIndex, dataset := range c.Datasets {
 		if strings.TrimSpace(dataset.Source) == "" {
-			return errors.New("dataset source is not defined")
+			return errors.Errorf("datasets[%d].source is not defined", datasetIndex)
 		}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go:453]: Valid finding. Follow-up PR #837, commit `6d85a68`, rejects leading or trailing dataset-source whitespace in both the Go frontend and Python defensive parser before URL/Hub classification. Errors are indexed and source-neutral, with credential/query/fragment redaction tests for SFT, DPO, and later dataset entries.
